src/bencode_parser.cpp

```diff
@@ -0,0 +1,191 @@
+  std::regex reg_string("^\\d+:$");
+  if (!_tokens.empty()) {
+    auto token = _tokens[_index];
+    if (token == "i") {
+      // If the first token is "i", then it's an integer, so we have a fixed
+      // length
+      if (_index == 0 && _tokens.size() != 3) {
+        throw std::invalid_argument(
+            "BencodeParser::decode(std::string decoded): invalid string");
+      }
+      return BencodeParser::decode_int();
+    } else if (std::regex_match(token, reg_string)) {
+      // If the first tokes is for a string, then it's a string, so we have a
+      // fixed length
+      if (_index == 0 && _tokens.size() != 2) {
+        throw std::invalid_argument(
+            "BencodeParser::decode(std::string decoded): invalid string");
+      }
+      return BencodeParser::decode_string();
+    } else if (token == "l") {
+      return BencodeParser::decode_list();
+    } else if (token == "d") {
+      return BencodeParser::decode_dict();
+    }
+  }
+  throw std::invalid_argument(
+      "BencodeParser::decode(std::string decoded): invalid string");
+}
+
+std::unique_ptr<BencodeValue> BencodeParser::decode_int() {
+  // The token must be in the form ["i", "number", "e"]
+  if (_tokens.size() - _index < 3) {
+    throw std::invalid_argument(
+        "BencodeParser::decode_int(std::vector<std::string>& encoded): no "
+        "tokens to decode");
+  }
+  if (_tokens[_index] != "i" || _tokens[_index + 2] != "e") {
+    throw std::invalid_argument(
+        "BencodeParser::decode_int(std::vector<std::string>& encoded): invalid "
+        "encoded string");
+  }
+  if (_tokens[_index + 1] == "-0") {
+    throw std::invalid_argument(
+        "BencodeParser::decode_int(std::vector<std::string>& encoded): "
+        "negative zero is not allowed");
+  }
+  // skip "e", "number" and "i" for the next decode
+  _index += 3;
+  return std::make_unique<BencodeInt>(std::stoi(_tokens[_index - 2]));
+}
+std::unique_ptr<BencodeValue> BencodeParser::decode_string() {
+  // The token must be in the form ["length:", "string"]
+  if (_tokens.size() - _index < 2) {
+    throw std::invalid_argument(
+        "BencodeParser::bencode_string(std::vector<std::string>& encoded): "
+        "invalid encoded string");
+  }
+  std::string str = _tokens[_index + 1];
+  // Transform n: to a integer
+  std::string length = _tokens[_index].substr(0, _tokens[_index].size() - 1);
+  int len = std::stoi(length);
+  // Check if the length of the string is correct
+  if (len < 0 || static_cast<unsigned int>(len) != str.size()) {
+    throw std::invalid_argument(
+        "BencodeParser::bencode_string(std::vector<std::string>& encoded): "
+        "invalid string length");
+  }
+  // "length:" and "string" for the next decode
+  _index += 2;
+  return std::make_unique<BencodeString>(str);
+}
+std::unique_ptr<BencodeValue> BencodeParser::decode_list() {
+  // The token must be in the form ["l",...,"e"]
+  if (_tokens.size() - _index < 4) {
```

Comment:
Shouldn't this need to be `< 2` to allow empty lists? The same observation can be made for `decode_dict`

Reply:
I had supposed that lists cannot be empty, but isn't specified on bencode specification